theano/gpuarray/opt.py

```diff
@@ -570,22 +854,29 @@ def local_gpu_pdbbreakpoint_op(node):
 
 @register_opt('fast_compile')
 @op_lifter([IfElse])
-def local_gpua_lazy_ifelse(node, context_name):
-    if node.op.gpu:
+@register_opt2([IfElse], 'fast_compile')
+def local_gpua_lazy_ifelse(op, context_name, inputs, outputs):
+    if op.gpu:
         return
-    c = node.inputs[0]
+    # this node is already on GPU, so don't change the graph
+    if isinstance(inputs[0].type, GpuArrayType):
+        return
```

Comment:
I understand this now. Currently(the code now) we're moving to the GPU only if any of the inputs are on the GPU. But @nouiz and me were having discussion as to how efficient the optimizer would be if we make this check or we don't and move the node entirely to the GPU(the aggressive mode). In that case, we would be trying to move those nodes that are already on the GPU back again and would create issue. That is why we had these checks on the nodes if they're on the GPU or not. 
Now, before this PR gets merged we must maintain the consistency in the code, 
1) We be aggressive and check this condition
2) We'd be strict and remove all the checking on the inputs of the node. 

Fred told, we should be aggressive when we merge and for the time being we're being strict.
https://github.com/Theano/Theano/pull/4570#discussion_r67357040